lib/client.js

```diff
@@ -78,6 +81,11 @@ module.exports = function (dependencies) {
               this.config.token.regenerate(tokenGeneration);
               resolve(this.write(notification, device, retryCount + 1));
               return;
+            } else if (status === "500" && response.reason === "InternalServerError") {
+              stream.connection.close();
+              let error = new VError("Error 500, stream ended unexpectedly");
```

Comment:
Here are my thoughts on this.

We currently return an error object if the connection fails, socket closes, etc. However, we don't return an error for HTTP status codes which could be considered an error like 403 or 500. I think we should keep it that way.

However, as I understand it, if the APNs server returns an HTTP 500 the connection should be considered unusable. Sending additional requests will fail. This is probably a bug in the APNs backend. So in that case it is OK to return an error object as a workaround and close the stream. This is actually what the implementation in this pull request does.

So, I think the current implementation in this PR is OK.